From 6fceeec53702635b696ff17a16e61199dba7e919 Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:18:49 -0400
Subject: [net] netfilter: nf_tables: use RCU-safe list insertion when replacing rules

Message-id: <20141021161849.22888.96524.stgit@localhost.localdomain>
Patchwork-id: 97965
O-Subject: [RHEL7 net PATCH 068/124] netfilter: nf_tables: use RCU-safe list insertion when replacing rules
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: 5bc5c307653cbf8fe9da6cbd8ae6c6bd5b86ff4b

 commit 5bc5c307653cbf8fe9da6cbd8ae6c6bd5b86ff4b
 Author: Pablo Neira Ayuso <pablo@netfilter.org>
 Date:   Tue Jun 10 10:53:00 2014 +0200

    netfilter: nf_tables: use RCU-safe list insertion when replacing rules

    The patch 5e94846 ("netfilter: nf_tables: add insert operation") did
    not include RCU-safe list insertion when replacing rules.

    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/netfilter/nf_tables_api.c b/net/netfilter/nf_tables_api.c
index 624e083..ba37c10 100644
--- a/net/netfilter/nf_tables_api.c
+++ b/net/netfilter/nf_tables_api.c
@@ -1796,7 +1796,7 @@ static int nf_tables_newrule(struct sock *nlsk, struct sk_buff *skb,
     goto err2;
    }
    nft_rule_disactivate_next(net, old_rule);
-   list_add_tail(&rule->list, &old_rule->list);
+   list_add_tail_rcu(&rule->list, &old_rule->list);
   } else {
    err = -ENOENT;
    goto err2;
-- 
1.7.1